staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/monitoring.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "crd"
+)
+
+var (
+	latencyBuckets = prometheus.ExponentialBuckets(1, 2, 5)
+)
+
+// converterMonitoringFactory holds metrics for all CRD converters
+type converterMonitoringFactory struct {
+	latencies map[string]*prometheus.HistogramVec
+}
+
+func newConverterMonitoringFactory() *converterMonitoringFactory {
+	return &converterMonitoringFactory{map[string]*prometheus.HistogramVec{}}
+}
+
+var _ crConverterInterface = &converterMonitoring{}
+
+type converterMonitoring struct {
+	delegate  crConverterInterface
+	latencies *prometheus.HistogramVec
+	labels    []string
+	crdName   string
+}
+
+func (c *converterMonitoringFactory) monitor(converterName string, crdName string, converter crConverterInterface) (crConverterInterface, error) {
+	metric, exists := c.latencies[converterName]
```

Comment:
is CRConverterFactory#NewConverter called on a single thread, or does this need to be threadsafe?

Reply:
I changed the code to use sync.Map but realized it is unnecessary and will make the code slower. this monitor system only being called when we call a NewConverter which only being called when a CRD storage is created which is guarded by a lock: 

https://github.com/kubernetes/kubernetes/blob/48adc8ec042edf930a972e01e959030f67a77023/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L390-L389

I think only documenting the method to say this is not thread-safe should suffice. wdyt?

Reply:
The other shared-state piece of the factory (NewWebhookConverter->HookClient->cm.cache.Get) is threadsafe. I'd make this map threadsafe as well so that parallel construction could work in the future.

Reply:
> this is not called often enough for lock contention to matter

Yep, I realized that after I commented. A simple mutex is even enough. We don't need to go sync.Map path.